### PR TITLE
[dtensor] fix requires_grad in distribute_tensor

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107608
* __->__ #107606

This PR fixes the requires_grad set when calling distribute_tensor, we
should set the requires_grad of the local tensor after the detach call
to make sure we create the leaf correctly, otherwise it would raise
warnings

### CODE REVIEW

**Review by fduwjj — APPROVED** [review]
LGTM
